base/src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java

```diff
@@ -0,0 +1,62 @@
+package uk.ac.ebi.atlas.experimentimport.sdrf;
+
+import org.springframework.util.StringUtils;
+import uk.ac.ebi.atlas.commons.readers.TsvStreamer;
+import uk.ac.ebi.atlas.resource.DataFileHub;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Named
+public class SdrfParser {
+
+    private final DataFileHub dataFileHub;
+
+    @Inject
+    public SdrfParser(DataFileHub dataFileHub) {
+        this.dataFileHub = dataFileHub;
+    }
+
+    /**
+     * Returns a map containing the header values for characteristics and factors, maintaining the same order in
+     * which they appear in the sdrf file.
+     */
+    public Map<String, Set<String>> parseHeader(String experimentAccession) {
+        try (TsvStreamer sdrfStreamer = dataFileHub.getExperimentFiles(experimentAccession).sdrf.get()) {
+            // Headers of interest are of the form HeaderType[HeaderValue]
+            Pattern pattern = Pattern.compile("(.*?)(\\[)(.*?)(].*)");
+
+            Map<String, Set<String>> headers = new HashMap<>();
+            Arrays.stream(
+                    sdrfStreamer
+                            .get()
+                            .findFirst()
+                            .orElse(new String[0]))
+                    .filter(x -> x.toLowerCase().contains("characteristic") || x.toLowerCase().contains("factor"))
+                    .forEach(header -> {
```

Comment:
This part makes me frown a bit... relying on side effects to populate a map is something I’d in principle discourage. Is there any part here that prevents you from e.g. mapping to a `Pair<String, Set<String>>` and then collect to a map? This case is not especially complex, but having a mapping `Function<>` means that you can also test it on its own if desired.

Reply:
What do you mean by relying on side effects? The fact that a `Map` has unique keys?

Reply:
I can't see how I can easily map to a `Pair<String, Set<String>`... I can go `String` --> `Pair<String, String>` --> `Map<String, Set<String>>` maybe... Though you won't always get a pair of something (if the matcher doesn't match), so I'd have to filter out the empty pairs... Unless I'm thinking of this in the completely wrong way.

Reply:
By side effects I mean that `forEach` takes a `Consumer` as an argument, which returns `void`, but the map is populated inside.

From [Wikipedia](https://en.wikipedia.org/wiki/Side_effect_(computer_science)):
> In computer science, a function or expression is said to have a side effect if it modifies some state outside its local environment, that is to say has an observable interaction with the outside world besides returning a value.

From [Java 8 API](https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html):
> Unlike most other functional interfaces, Consumer is expected to operate via side-effects.

Reply:
In essence I can see a function that takes a stream of headers and creates an entry per header if they match a condition. To me that adjusts to a filter-map pattern:
```
String regex = "(.*?)(\\[)(.*?)(].*)";
Pattern pattern = Pattern.compile("(.*?)(\\[)(.*?)(].*)");
...
Arrays.stream(...)
    .filter(x -> x.toLowerCase().contains("characteristic") || x.toLowerCase().contains("factor"))
    .filter(x -> x.matches(regex))
    .map(x -> // return a pair)
    .collect(toMap(Pair::getLeft, Pair::getRight))
```

Reply:
If you rely on the map being partially populated, you can use a `toMap` method with a third argument to merge values: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toMap-java.util.function.Function-java.util.function.Function-java.util.function.BinaryOperator-

Reply:
Another option is to collect using `groupingBy` at the end if you don’t mind ending with a `Map<K, List<V>>`.

Reply:
Beautiful! :) Pure functions, wherever possible, are the way to go.